README.md
# kotatsu

**kotatsu** is a straightforward CLI tool aiming either at running node.js scripts or serving JavaScript/TypeScript web applications in a modern environment (modules, ES2015, [Hot Module Replacement](https://webpack.github.io/docs/hot-module-replacement.html), etc.).

Its goal is to relieve developers from the really heavy stack that we now face on a daily basis when working with modern JavaScript.

The idea is to let developers new to the stack forget about it as long as they can while enabling seasoned developers to setup their environment very fast and to start customizing the stack progressively only when this is really needed.

Typical use cases for **kotatsu** are hot-reloaded [express](http://expressjs.com/) APIs written in ES2015, hot-reloaded [React](https://facebook.github.io/react/) or [deku](http://dekujs.github.io/deku/) applications etc. Check the [use cases](#use-cases) for a quick glance of what can be achieved.

Note that **kotatsu** currently uses [webpack](https://webpack.github.io/docs/) under the hood to perform its magic.

## Summary

* [Installation](#installation)
* [Usage](#usage)
* [Package Config](#package-config)
* [Typescript](#typescript)
* [Style](#style)
* [Use Cases](#use-cases)
  * [Interval](#interval)
  * [Express](#express)
  * [Deku](#deku)
  * [React](#react)
* [Node API](#node-api)
  * [start](#start)
  * [serve](#serve)
  * [monitor](#monitor)
  * [run](#run)
  * [build](#build)
* [What on earth is a kotatsu?](#explanation)
* [Inspiration](#inspiration)
* [License](#license)

## Installation

Kotatsu can be installed globally (you should avoid this!) or within your node.js project using npm:

```bash
# Within your project
npm install --save-dev kotatsu

# Globally:
[sudo] npm install -g kotatsu
```

## Usage

```
Usage: kotatsu <command> {options} [entry]

Commands:
  cli.js start    Start a node.js script.
  cli.js serve    Serve a client-side application.
  cli.js monitor  Monitor a terminating node.js script.
  cli.js run      Run the given node.js script.
  cli.js build    Build your code for client or server.

Options:
  -c, --config            Optional webpack config that will be merged with kotatsu's one (useful if
                          you need specific loaders).
  -d, --devtool           Webpack devtool spec to use to compute source maps.               [string]
  -m, --mount-node        Id of the mount node in the generated HMTL index.                 [string]
  -o, --output            Output path (either directory or filename).                       [string]
  -p, --port              Port that the server should listen to.                     [default: 3000]
  -s, --source-maps       Should source maps be computed for easier debugging?
                                                                           [boolean] [default: true]
      --cors              Should the server allow CORS?                    [boolean] [default: true]
      --index             Path to a custom HMTL index file. Will default to `./index.html` if
                          present.                                                          [string]
      --jsx               Does your code uses JSX syntax?                 [boolean] [default: false]
      --pragma            JSX pragma.                                                       [string]
      --sass              Whether to transpile scss files (requires `sass` or `node-sass`).
                                                                          [boolean] [default: false]
      --typescript, --ts  Whether to support TypeScript (requires `typescript`).
                                                                          [boolean] [default: false]
      --open              Whether to open your browser on the served application.
                                                                          [boolean] [default: false]
      --presets           Babel presets separated by a comma (example:
                          @babel/preset-stage-2,@babel/preset-react).                       [string]
      --production        Whether to build for production (minify + define).
                                                                          [boolean] [default: false]
      --progress          Should it display the compilation's progress?   [boolean] [default: false]
      --proxy             Proxy information (example: --proxy /api http://localhost:4000)   [string]
      --public            Mounting a path to a public folder (example: --public /data ./src/data).
                          Can be used several times.                                        [string]
      --quiet             Disable logs.                                   [boolean] [default: false]
      --version           Show version number                                              [boolean]
  -h, --help              Show help                                                        [boolean]

Examples:
  kotatsu start script.js                           Launching the given script with HMR.
  kotatsu start -c webpack.config.js script.js      Using a specific webpack config.
  kotatsu start --no-source-maps script.js          Disabling source maps.
  kotatsu start script.js -- --path test.js         Passing arguments to the script.

  kotatsu serve entry.js                            Serving the given app.
  kotatsu serve --jsx entry.jsx                     Serving the given app with JSX code.
  kotatsu serve --port 8000 entry.jsx               Serving the app on a different port.
  kotatsu serve --proxy /api http://localhost:4000  Proxying an API.
  kotatsu serve --public /data ./src/data           Serving local static files.
  kotatsu serve --sass entry.js                     Supporting SASS stylesheets.
  kotatsu serve --typescript entry.ts               Serving a TypeScript app.

  kotatsu build server entry.js -o ./               Build the given server script.
  kotatsu build client --production entry.js -o ./  Build the given client app for production.
```

If this is your first time using **kotatsu**, you should really check the use cases below to see how it could fit your workflow.

## Package config

If you don't want to repeat yourself in your npm scripts know that you can use the `kotatsu` key of your project's `package.json` to keep a global configuration for the CLI:

```json
{
  "scripts": {
    "dev": "kotatsu serve ./entry.js"
  },
  "kotatsu": {
    "progress": true,
    "public": ["/data", "./src/data"]
  }
}
```

## TypeScript

**kotatsu** supports TypeScript out of the box. If you want to use TypeScript in your project, just ensure you have a valid `tsconfig.json` file and that you have installed the `typescript` dependencies. Then just use the `--typescript` flag and everything should work just fine.

## Style

**kotatsu** lets you import CSS files out of the box. If you need to import SCSS files, you can use the `--sass` flag but be sure to install `sass` (or `node-sass`) for it to work.

## Use cases

### Interval

This example does not really serve a real-life purpose but merely shows you how to hot-reload a very simple node.js script.

The idea here is to create a script that will continuously print a required string into the console every 2 seconds:

**1. Creating the necessary files**

```js
// file: interval.js
var string = require('./string.js');

setInterval(function() {
  console.log(string);
}, 2000);

// Here is the twist: whenever the `string` dependency is updated, we will swap it:
if (module.hot) {
  module.hot.accept('./string.js', function() {
    string = require('./string.js');
  });
}
```

```js
// file: string.js
module.exports = 'ping';
```

**2. Using kotatsu to start the script**

```bash
kotatsu start interval.js
```

Now the script will start and you should see it logging `ping` into the console every two seconds.

Now edit the `string.js` file and the script will automatically update and log the new exported value of the file.

**3. Let's use the same script in the browser**

You would rather run this script in the browser?

```js
kotatsu serve interval.js
```

Now go to `localhost:3000` and you should be able to observe the same kind of results in the console.

**Remarks**

This example serves another purposes: showing you that **kotatsu** is meant to be used on long-running scripts such as servers or UIs. If what you need is to code a terminating script, check the `monitor` command instead.

If you need more information about `module.hot` and Hot Module Replacement (HMR), go check webpack's [docs](https://webpack.github.io/docs/hot-module-replacement.html) on the subject.

### Express

Let's setup a very simple hot-reloaded express app:

**1. Installing necessary dependencies**

```bash
npm i --save express
npm i --save-dev kotatsu
```

**2. Creating our app**

```js
// file: app.js
var express = require('express');

var app = express();

app.get('/', function(req, res) {
  return res.send('Hello World!');
});

module.exports = app;
```

**3. Creating our startup script**

```js
// file: start.js
var app = require('./app.js'),
    http = require('http');

var server = http.createServer(app);

server.listen(3000);

if (module.hot) {

  // This will handle HMR and reload the server
  module.hot.accept('./app.js', function() {
    server.removeListener('request', app);
    app = require('./app.js');
    server.on('request', app);
    console.log('Server reloaded!');
  });
}
```

**4. Using kotatsu**

Launching our app with HMR so we can work comfortably.

```bash
kotatsu start ./start.js
```

You can now edit the express app live and it will automatically update without having to reload the script.

### Deku

**1. Installing necessary dependencies**

```bash
npm i --save deku
npm i --save-dev kotatsu
```

**2. Creating our main component**

```jsx
// file: App.jsx
import {element} from 'deku';

export default function App() {
  return <div>Hello World!</div>;
}
```

**3. Creating our application's entry**

```jsx
// file: main.jsx
import {dom, element} from 'deku';
import InitalApp from './App.jsx';

const mountNode = document.getElementById('app'),
      render = dom.createRenderer(mountNode);

function refresh(Component) {
  render(<Component />);
}
refresh(InitalApp);

// Let's handle our code's updates
if (module.hot) {
  module.hot.accept('./App.jsx', function() {
    const NextApp = require('./App.jsx');
    refresh(NextApp);
  });
}
```

**4. Using kotatsu**

Now let's run a server to host our app:

```bash
kotatsu serve --jsx --pragma element main.jsx
```

Note that **kotatsu** will serve for you a HTML index file looking quite like this:

```html
<!DOCTYPE html>
<html>
  <head>
    <title>kotatsu</title>
  </head>
  <body>
    <div id="app"></div>
    <script type="text/javascript" src="/build/bundle.js"></script>
  </body>
</html>
```

If you need a custom one, just use the `--index` argument.

Now visit `localhost:3000` and you are ready to develop.

### React

React is a bit more tricky because we need to install a Babel 6 preset (`react-hmre`) to handle hot-loading.

**1. Installing necessary dependencies**

```bash
npm i --save react react-dom
npm i --save-dev kotatsu babel-preset-react babel-preset-react-hmre
```

**2. Creating our main component**

```jsx
// file: App.jsx
import React, {Component} from 'react';

export default class App extends Component {
  render() {
    return <div>Hello World!</div>;
  }
}
```

**3. Creating our application's entry**

```jsx
// file: main.jsx
import React from 'react';
import {render} from 'react-dom';
import App from './App.jsx';

const mountNode = document.getElementById('app');

render(<App />, mountNode);
```

**4. Using kotatsu**

Now let's run a server to host our app:

```bash
kotatsu serve --jsx ./main.jsx
```

Note that **kotatsu** will serve for you a HTML index file looking quite like this:

```html
<!DOCTYPE html>
<html>
  <head>
    <title>kotatsu</title>
  </head>
  <body>
    <div id="app"></div>
    <script type="text/javascript" src="/build/bundle.js"></script>
  </body>
</html>
```

If you need a custom one, just use the `--index` argument.

Now visit `localhost:3000` and you are ready to develop.

## Node API

```js
var kotatsu = require('kotatsu');
```

Every method of the library uses the same configuration object (similar to the CLI arguments):

*required*

* **entry** *string*: Path towards the entry.

*optional*

* **args** *array*: array of arguments to pass to the child script.
* **cwd** *string* [`process.cwd()`]: current working directory.
* **config** *object*: a webpack config object.
* **cors** *boolean* [`true`]: should the server allow CORS?
* **devtool** *string*: a webpack devtool [spec](https://webpack.github.io/docs/configuration.html#devtool).
* **index** *string*: path of the HTML index file to serve.
* **jsx** *boolean* [`false`]: should we handle JSX?
* **mountNode** *string* [`'app'`]: id of the mount node in the generated HTML index file.
* **open** *boolean* [`false`]: Whether to open the app in you web browser as soon as served.
* **output** *string* [`.kotatsu`]: path of the built file.
* **port** *integer* [`3000`]: port that the server should listen to.
* **pragma** *string*: custom JSX pragma.
* **presets** *array*: Babel 6 presets to apply.
* **production** *boolan*: Whether to build for production, i.e. minify output and define `NODE_ENV` as `production`.
* **progress** *boolean* [`false`]: should the compiler display a progress bar?
* **proxy** *array*: proxy information.
* **quiet** *boolean* [`false`]: if true, will disable all console logs.
* **server** *function*: function called with the express app in case you want to apply custom middlewares etc.
* **sourceMaps** *boolean* [`true`]: should it compute source maps?

### start

```js
var watcher = kotatsu.start({
  entry: 'script.js',
  ...
})
```

### serve

```js
var server = kotatsu.serve({
  entry: 'script.js',
  ...
})
```

### monitor

```js
var watcher = kotatsu.monitor({
  entry: 'script.js',
  ...
})
```

### run

```js
kotatsu.run({
  entry: 'script.js',
  ...
});
```

### build

```js
// side can be either 'front' or 'back'
kotatsu.build(side, {
  entry: 'script.js',
  ...
}, callback);
```

<h2 id="explanation">What on earth is a kotatsu?</h2>

A [kotatsu](https://en.wikipedia.org/wiki/Kotatsu) is a low Japanase table covered by a heavy blanket with an underneath heat source that keeps you warm in the cold season.

## Inspiration

**kotatsu** is widely inspired by the following modules:

- [webpack-hot-middleware](https://github.com/glenjamin/webpack-hot-middleware) by [@glenjamin](https://github.com/glenjamin).
- [nodemon](https://github.com/remy/nodemon) by [@remy](https://github.com/remy).

## License

[MIT](LICENSE.txt)


CHANGELOG.md
# Changelog

## 0.22.3

* Fixing callback not being called on success with `kotatsu.build`.

## 0.22.2

* Upgrading deps.

## 0.22.1

* Upgrading deps.

## 0.22.0

* Adding support for `resolve.extensions`.
* Adding the `--open` flag.
* Detecting the presence of `index.html` in cwd for convenience.
* Dropping obsolete `--babel` flag.
* Fixing ties to webpack v5.

## 0.21.0

* Adapting `kotatsu` to webpack v5.

## 0.20.0

* `--public` is now nargs=2 so you can specify a custom mounting point.
* Adding `package.json` configuration through the `kotatsu` key.

## 0.18.0

* Upgrading dependencies.

## 0.17.0

* Handling object rest spread and class properties by default.
* Fixing default build path for client.
* Upgrading dependencies.

## 0.16.0

* Upgrading all dependencies (latest babel & webpack).
* Transpiling `es2015` code by default.
* Adding callback to `build`.
* Changing `--minify` into `--production`.
* CSS support.
* SCSS support.

## 0.15.2

* Fixing `node_modules` handling for backend usage.

## 0.15.1

* Upgrading to `babel-loader` 7.0.0.

## 0.15.0

* Upgrading dependencies and switching to webpack 2.

## 0.14.0

* Adding unicode support for default HTML ([@bverjat](https://github.com/bverjat)).

## 0.13.0

* Adding the `cors` option ([@SeeThruHead](https://github.com/SeeThruHead)).

## 0.12.2

* Fixing issues concerning `require.resolve` on Windows.
* Exiting with the correct code when failing a build ([@SeeThruHead](https://github.com/SeeThruHead)).

## 0.12.1

* Fixing an issue related to child processes with Windows.

## 0.12.0

* Redirecting every route on the index.
* Adding the `server` option.
* Fixing an issue concerning default options.
* Updating `http-proxy-middleware`.

## 0.11.0

* Updating dependencies.
* Proxy is now copying headers correctly.
* Source maps are now default.
* Merging some specific entries to avoid collisions.

## 0.10.0

* Adding plugins merging.
* Fixing ES6 config support.

## 0.9.1

* Cleaner log for the `serve` command.
* Better log for the `build` command.
* Fixing config merging precedences.

## 0.9.0

* Handling `webpack.config.babel.js`.

## 0.8.3

* Attempting to resolve entry file rather than checking the file's existence.
* Better priorities when solving webpack config.
* Fixing update logs.

## 0.8.2

* Fixing single string as webpack entry.
* Fixing error display when failing to load webpack config.

## 0.8.1

* Fixing missing dependency.

## 0.8.0

* Adding `--proxy`.
* The `devServer` options of the webpack config are now passed to `webpack-dev-middleware`.
* Possibility to read the entry from a webpack config.
* Fixing an issue with the `run` command watching for changes.

## 0.7.0

* Adding the `run` command.
* Adding the `build` command.
* Adding the `monitor` command.
* Adding the `--minify` option.
* Better `--output` option.
* Improving UX.

## 0.6.0

* Adding `--public`.
* Back to lodash v3 for faster installation.

## 0.5.0

* Adding `--babel`.

## 0.4.2

* Using `--devtool` now implicitly enables `--source-maps`.

## 0.4.1

* Handling warnings correctly.

## 0.4.0

* Adding json support for client.
* Fixing a progress bar collision.

## 0.3.3

* Fixing a bug when using both `--es2015` and `--presets`.
* Adding `NoErrorsPlugin` back.

## 0.3.2

* Fixing `__dirname` and `__filename`.

## 0.3.1

* Fixing json support for node.

## 0.3.0

* Adding `--output`.
* Adding `--presets`.
* Better logs.
* Supporting script arguments through `--`.
* Fixing `--es2015` and `--jsx`.

## 0.2.0

* New tool covering client-side usage.

## 0.1.2

* Fixing yet another issue with JSON loader.

## 0.1.1

* Fixing an issue with the JSON loader.

## 0.1.0

* Handling compilation errors.
* Handling relative paths without `./` as entry.
* Handling JSON out of the box.

## 0.0.8

* Adding `--version`.
* Fixing and issue with `__dirname` and `__filename`.


cli.js
#!/usr/bin/env node
/* eslint no-shadow: 0 */
/**
 * Kotatsu CLI
 * ============
 *
 * The CLI tool that will call the lib's function.
 */
var kotatsu = require('./kotatsu.js'),
    red = require('chalk').red,
    yargs = require('yargs'),
    path = require('path'),
    pkg = require('./package.json'),
    fs = require('fs'),
    _ = require('lodash');

// Handling ES6 configuration
require('@babel/register')({
  only: [/\.babel\.js/],
  presets: ['@babel/env']
});

var CWD = process.cwd();

var COMMANDS = [
  'start',
  'serve',
  'monitor',
  'run',
  'build'
];

var EXPECTED_PARTS = 2;

var USAGE = 'Usage: kotatsu <command> {options} [entry]',
    AVAILABLE_COMMANDS = 'Available commands: ' + COMMANDS.join(', ');

function error(message, details) {
  throw Error('\n' + red('Error: ' + message) + (details ? '\n' + details : ''));
}

var webpackConfig = {},
    entry = null,
    command,
    side;

// Building the CLI
var argv = yargs
  .locale('en')
  .wrap(100)
  .usage(USAGE)
  .pkgConf('kotatsu')
  .check(function(argv) {
    command = argv._[0];

    if (!argv._.length)
      error('Not enough arguments.', [
        USAGE,
        AVAILABLE_COMMANDS
      ].join('\n'));

    if (!Number.isInteger(argv.port))
      error('Invalid port: ' + argv.port);

    if (!~COMMANDS.indexOf(command))
      error(
        'Invalid command: ' + command,
        AVAILABLE_COMMANDS
      );

    // Should we load a config file?
    if (argv.config) {
      try {
        var configPath = require.resolve(path.join(process.cwd(), argv.config));
        webpackConfig = require(configPath);

        // Handling ES6 export
        if ('default' in webpackConfig)
          webpackConfig = webpackConfig.default;
      }
      catch (e) {
        error(
          'Error while loading your config file: "' + argv.config + '".',
          e.stack || ''
        );
      }

      if (webpackConfig.entry)
        EXPECTED_PARTS--;
    }

    // Build specifics
    if (command === 'build') {
      EXPECTED_PARTS++;

      if (argv._.length < EXPECTED_PARTS)
        error('The "build" command takes 2 arguments: client or server, and the entry.');

      if (!~['client', 'server'].indexOf(argv._[1]))
        error('Do you want to build for client or server? You gave: "' + argv._[1] + '".');

      side = argv._[1] === 'client' ? 'front' : 'back';
      entry = argv._[2];
    }
    else {
      if (argv._.length < EXPECTED_PARTS)
        error('Expecting two arguments: the command and the path to your entry.');

      entry = argv._[1];
    }

    // Attempting to resolve our entry
    if (entry) {
      entry = path.resolve(CWD, entry);

      try {
        require.resolve(entry);
      }
      catch (e) {
        error('Could not resolve your entry: "' + argv._[1] + '".', e.stack);
      }
    }

    return true;
  })

  // Commands
  .command('start', 'Start a node.js script.')
  .command('serve', 'Serve a client-side application.')
  .command('monitor', 'Monitor a terminating node.js script.')
  .command('run', 'Run the given node.js script.')
  .command('build', 'Build your code for client or server.')

  // Generic options
  .option('c', {
    alias: 'config',
    describe: 'Optional webpack config that will be merged with kotatsu\'s one (useful if you need specific loaders).'
  })
  .option('d', {
    alias: 'devtool',
    describe: 'Webpack devtool spec to use to compute source maps.',
    type: 'string'
  })
  .option('m', {
    alias: 'mount-node',
    describe: 'Id of the mount node in the generated HMTL index.',
    type: 'string'
  })
  .option('o', {
    alias: 'output',
    describe: 'Output path (either directory or filename).',
    type: 'string'
  })
  .option('p', {
    alias: 'port',
    describe: 'Port that the server should listen to.',
    default: 3000
  })
  .option('s', {
    alias: 'source-maps',
    describe: 'Should source maps be computed for easier debugging?',
    type: 'boolean',
    default: true
  })
  .option('cors', {
    describe: 'Should the server allow CORS?',
    type: 'boolean',
    default: true
  })
  .option('index', {
    describe: 'Path to a custom HMTL index file. Will default to `./index.html` if present.',
    type: 'string'
  })
  .option('jsx', {
    describe: 'Does your code uses JSX syntax?',
    type: 'boolean',
    default: false
  })
  .option('pragma', {
    describe: 'JSX pragma.',
    type: 'string'
  })
  .option('sass', {
    describe: 'Whether to transpile scss files (requires `sass` or `node-sass`).',
    type: 'boolean',
    default: false
  })
  .option('typescript', {
    alias: 'ts',
    describe: 'Whether to support TypeScript (requires `typescript`).',
    type: 'boolean',
    default: false
  })
  .option('open', {
    describe: 'Whether to open your browser on the served application.',
    type: 'boolean',
    default: false
  })
  .option('presets', {
    describe: 'Babel presets separated by a comma (example: @babel/preset-stage-2,@babel/preset-react).',
    type: 'string'
  })
  .option('production', {
    describe: 'Whether to build for production (minify + define).',
    type: 'boolean',
    default: false
  })
  .option('progress', {
    describe: 'Should it display the compilation\'s progress?',
    type: 'boolean',
    default: false
  })
  .option('proxy', {
    describe: 'Proxy information (example: --proxy /api http://localhost:4000)',
    type: 'string',
    nargs: 2
  })
  .option('public', {
    describe: 'Mounting a path to a public folder (example: --public /data ./src/data). Can be used several times.',
    type: 'string',
    nargs: 2
  })
  .option('quiet', {
    describe: 'Disable logs.',
    type: 'boolean',
    default: false
  })

  // Examples
  .example('kotatsu start script.js', 'Launching the given script with HMR.')
  .example('kotatsu start -c webpack.config.js script.js', 'Using a specific webpack config.')
  .example('kotatsu start --no-source-maps script.js', 'Disabling source maps.')
  .example('kotatsu start script.js -- --path test.js', 'Passing arguments to the script.')
  .example('')
  .example('kotatsu serve entry.js', 'Serving the given app.')
  .example('kotatsu serve --jsx entry.jsx', 'Serving the given app with JSX code.')
  .example('kotatsu serve --port 8000 entry.jsx', 'Serving the app on a different port.')
  .example('kotatsu serve --proxy /api http://localhost:4000', 'Proxying an API.')
  .example('kotatsu serve --public /data ./src/data', 'Serving local static files.')
  .example('kotatsu serve --sass entry.js', 'Supporting SASS stylesheets.')
  .example('kotatsu serve --typescript entry.ts', 'Serving a TypeScript app.')
  .example('')
  .example('kotatsu build server entry.js -o ./', 'Build the given server script.')
  .example('kotatsu build client --production entry.js -o ./', 'Build the given client app for production.')

  // Help & Version
  .version(pkg.version)
  .help('h')
  .alias('h', 'help')
  .showHelpOnFail(false, 'Use the --help option to get the list of available options.')
  .epilogue('Repository: ' + pkg.repository.url)
  .argv;

var publicPaths = argv.public;

if (publicPaths) {
  publicPaths = _.chunk(publicPaths, 2).map(function(p) {
    return [p[0], path.resolve(CWD, p[1])];
  });
}

if (!argv.index && fs.existsSync(path.join(CWD, 'index.html')))
  argv.index = 'index.html';

var opts = {
  args: argv._.slice(EXPECTED_PARTS),
  cors: argv.cors,
  cwd: CWD,
  config: webpackConfig,
  devtool: argv.devtool,
  entry: entry ? path.resolve(CWD, entry) : null,
  index: argv.index ? path.resolve(CWD, argv.index) : null,
  public: publicPaths,
  jsx: argv.jsx,
  mountNode: argv.mountNode,
  open: argv.open,
  output: argv.output ? path.resolve(CWD, argv.output) : null,
  port: argv.port,
  pragma: argv.pragma,
  presets: argv.presets ? argv.presets.split(',') : null,
  production: argv.production,
  progress: argv.progress,
  proxy: argv.proxy,
  quiet: argv.quiet,
  sass: argv.sass,
  sourceMaps: argv.sourceMaps,
  typescript: argv.typescript
};

// Cleaning null values
for (var k in opts)
  if (opts[k] === null || opts[k] === undefined)
    delete opts[k];

// Applying the correct method.
if (command === 'build') {
  kotatsu.build(side, opts, function(err) {
    if (err) {
      console.error(err);
      process.exit(1);
    }
  });
}
else {
  kotatsu[command](opts);
}
